rfcs/021-logging/README.md

```diff
@@ -0,0 +1,113 @@
+
+AWS provides pre-built images for running both Fluent Bit and Fluentd in ECS and provide [instructions for setting up custom log routing](https://docs.aws.amazon.com/AmazonECS/latest/userguide/using_firelens.html).
+
+Firelens requires you to add a container to your task definition, though that could be provided with fluent bit configuration from the existing [`terraform-aws-ecs-service`](https://github.com/wellcomecollection/terraform-aws-ecs-service) module allowing us to standardise that configuration.
+
+We may still need to provide a mechanism to forward logs collected to Elasticsearch. 
+
+### Fluent Bit
+
+AWS recommends using Fluent Bit as a log collector over Fluentd [due to performance and resource considerations](https://aws.amazon.com/blogs/opensource/centralized-container-logging-fluent-bit/). 
+
+Fluent Bit is similar to Logstash in function and is capable of [relaying logs to Logstash](https://github.com/fluent/fluent-bit-tutorials/blob/master/ship_to/logstash.md) if required. See the [Fluent Bit](https://docs.fluentbit.io/manual/) manual for details of input & output plugins.
+
+### Routing logs to Elasticsearch
+
+At present we route logs through a cluster-local Logstash service that is configured to authenticate with a Elastic Cloud hosted Elasticsearch instance and forward logs there with several `logstash-transit` services across the estate.
+
+There are a few options for routing to Elasticsearch.
+
+- Directly to Elasticsearch from the Fluent Bit sidecar.
+  - (pro) No other log forwarding infrastructure.
+  - (con) Elasticsearch config would be required in each service.
+- Forwarding logs to a cluster-local collector as we do now.
+  - (???) We can use the Fluent Bit [`forward`](https://docs.fluentbit.io/manual/input/forward) input and so replace Logstash with Fluent Bit which _might_ be more reliable.
+  - (con) Multiple `???-transit` services would still be required.
+- Provide a single collector service across the platform utilising [AWS PrivateLink](https://docs.aws.amazon.com/vpc/latest/userguide/endpoint-service.html).
+  
+  VPC Endpoint Services (AWS PrivateLink) allows you to share an [NLB (Network Load Balancer)](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html) fronted service between VPCs and between AWS accounts. We have made use of AWS PrivateLink to share on-premises VPN accessible services (e.g. our Workflow service "Goobi") between accounts with success in the past.
+  
+  Endpoint services would be provisioned alongside network and account configuration in the [platform-infrastructure](https://github.com/wellcomecollection/platform-infrastructure) terraform stack. 
+  
+  - (pro) One service with central configuration for Elasticsearch authentication.
+  - (pro) We have experience with successfully using PrivateLink.
+  - (con) Requires some log collector infrastructure and network configuration though centralised and with a single instance of the collector service.
```

Comment:
Not sure if it's a pro or a con here, but the fact that if it goes down, all our logging goes down.

Would there still need to be a sidecar that talks to this service from each of the services?

Reply:
Just saw the diagram - makes more sense.

Reply:
So the "log collector endpoint" in the service account (not platform) is an ALB? Not sure I grasp what those _are_, but get what they're _doing_.

Reply:
> Not sure if it's a pro or a con here, but the fact that if it goes down, all our logging goes down.

Some truth to that, though I think we can mitigate that risk as we do in our other services (multiple tasks/cross AZ distribution).

> Would there still need to be a sidecar that talks to this service from each of the services?

This is described in the PR - the "Firelens" logging adapter feeds logs to a Fluent Bit sidecar, though AWS takes care of providing that image and we can package it so that it is transparent to consumers. I read somewhere that this is described as a "hidecar"!

Reply:
Because of the wordplay, I approve whole-heartedly.